quinn-h3/examples/h3.rs

```diff
@@ -0,0 +1,258 @@
+use std::net::SocketAddr;
+use std::net::ToSocketAddrs;
+use std::path::PathBuf;
+use std::rc::Rc;
+use std::sync::Arc;
+use std::time::Instant;
+use std::{fmt, fs, io};
+
+use failure::{bail, format_err, Error, Fail, ResultExt};
+use futures::{Future, Stream};
+use slog::{info, o, Drain, Logger};
+use structopt::{self, StructOpt};
+use tokio::runtime::current_thread::Runtime;
+use url::Url;
+
+use quinn::ConnectionDriver as QuicDriver;
+use quinn_h3::{
+    client::ClientBuilder,
+    connection::ConnectionDriver,
+    server::{IncomingRequest, ServerBuilder},
+};
+
+type Result<T> = std::result::Result<T, Error>;
+
+pub struct PrettyErr<'a>(&'a dyn Fail);
+impl<'a> fmt::Display for PrettyErr<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        fmt::Display::fmt(&self.0, f)?;
+        let mut x: &dyn Fail = self.0;
+        while let Some(cause) = x.cause() {
+            f.write_str(": ")?;
+            fmt::Display::fmt(&cause, f)?;
+            x = cause;
+        }
+        Ok(())
+    }
+}
+
+pub trait ErrorExt {
+    fn pretty(&self) -> PrettyErr<'_>;
+}
+
+impl ErrorExt for Error {
+    fn pretty(&self) -> PrettyErr<'_> {
+        PrettyErr(self.as_fail())
+    }
+}
+
+#[derive(StructOpt, Debug, Clone)]
+#[structopt(name = "h3")]
+struct Opt {
+    #[structopt(short = "s", long = "server")]
+    server: bool,
+    #[structopt(default_value = "http://127.0.0.1:4433/Cargo.toml")]
+    url: Url,
+    /// directory to serve files from
+    #[structopt(parse(from_os_str), default_value = ".")]
+    root: PathBuf,
+    /// TLS private key in PEM format
+    #[structopt(parse(from_os_str), short = "k", long = "key", requires = "cert")]
+    key: Option<PathBuf>,
+    /// TLS certificate in PEM format
+    #[structopt(parse(from_os_str), short = "c", long = "cert", requires = "key")]
+    cert: Option<PathBuf>,
+    /// Enable stateless retries
+    #[structopt(long = "stateless-retry")]
+    stateless_retry: bool,
+    /// Address to listen on
+    #[structopt(long = "listen", default_value = "0.0.0.0:4433")]
+    listen: SocketAddr,
+    /// Custom certificate authority to trust, in DER format
+    #[structopt(parse(from_os_str), long = "ca")]
+    ca: Option<PathBuf>,
+    /// Simulate NAT rebinding after connecting
+    #[structopt(long = "rebind")]
+    rebind: bool,
+}
+
+fn main() {
+    let opt = Opt::from_args();
+    let sdecorator = slog_term::PlainSyncDecorator::new(std::io::stderr());
+    let cdecorator = slog_term::PlainSyncDecorator::new(std::io::stderr());
```

Comment:
You shouldn't construct completely separate root loggers. Instead, construct a single one, then fork it with `log.new(o!(whatever))` as necessary.

Reply:
Thanks!